drivers/timer/mchp_xec_rtos_timer.c

```diff
@@ -0,0 +1,271 @@
+
+static struct k_spinlock lock;
+
+static volatile u32_t last_load;
+
+static volatile u32_t cycle_count;
+
+static volatile u32_t announced_cycles;
+
+static volatile u32_t elapsed_counts;
+
+static ALWAYS_INLINE void timer_clear_intr(void)
+{
+	GIRQ23_REGS->SRC = MCHP_RTMR_GIRQ_VAL;
+	GIRQ23_REGS->RESULT = GIRQ23_REGS->SRC;
+	NVIC_ClearPendingIRQ(RTMR_IRQn);
+}
+
+static ALWAYS_INLINE void timer_preload_set(u32_t val)
+{
+	RTMR_REGS->PRLD = val & COUNTER_MAX;
+}
+
+static ALWAYS_INLINE void timer_counter_set(u32_t val)
+{
+	RTMR_REGS->CNT = val & COUNTER_MAX;
+}
+
+static ALWAYS_INLINE void timer_start(void)
+{
+	RTMR_REGS->CTRL = MCHP_RTMR_CTRL_BLK_EN + MCHP_RTMR_CTRL_HW_HALT_EN
+			  + MCHP_RTMR_CTRL_START;
+}
+
+/*
+ * Side-effect all RTOS timer registers are cleared.
+ */
+static ALWAYS_INLINE void timer_stop(void)
+{
+	RTMR_REGS->CTRL = 0;
+}
+
+static ALWAYS_INLINE void timer_pause(void)
+{
+	RTMR_REGS->CTRL |= MCHP_RTMR_CTRL_FW_HALT_EN;
+}
+
+static ALWAYS_INLINE void timer_unpause(void)
+{
+	RTMR_REGS->CTRL &= ~(MCHP_RTMR_CTRL_FW_HALT_EN);
+}
+
+static INLINE u32_t timer_is_paused(void)
+{
+	return (RTMR_REGS->CTRL & MCHP_RTMR_CTRL_FW_HALT_EN);
+}
+
+static void timer_ien(void)
+{
+	GIRQ23_REGS->EN_SET = MCHP_RTMR_GIRQ_VAL;
+}
+
+/* SYSTICK elapsed function */
+#if 0
+static u32_t elapsed(void)
+{
+	u32_t val, ctrl1, ctrl2;
+
+	ctrl1 = GIRQ23_REGS->SRC;
+	val = RTMR_REGS->CNT & COUNTER_MAX;
+	ctrl2 = GIRQ23_REGS->SRC;
+
+	overflow_cyc += (ctrl1 & MCHP_RTMR_GIRQ_VAL) ? last_load : 0;
+	if (val > VAL_ABOUT_TO_WRAP) {
+		int wrap = ctrl2 & MCHP_RTMR_GIRQ_VAL;
+
+		overflow_cyc += (wrap != 0) ? last_load : 0;
+	}
+
+	return (last_load - val) + overflow_cyc;
+}
+#endif
+
+/*
+ * RTOS Timer runs at 32KHz and counts down to 0. The number of counts
+ * elpapsed since the timer was started is Preload register - Count register.
+ * This routine is expected to return the number of ticks elapsed since the
+ * last time this routine was called. This implementation depends upon
+ * one-shot behaviour (auto-reload NOT enabled).
+ */
+static u32_t elapsed(void)
+{
+	u32_t tdiff = (last_load - (RTMR_REGS->CNT & COUNTER_MAX));
+	u32_t elapsed = 0;
+
+	if (tdiff != elapsed_counts) {
+		elapsed_counts = tdiff;
+		elapsed = tdiff;
+	}
+	return elapsed;
+}
+
+void xec_rtos_timer_isr(void *arg)
+{
+	ARG_UNUSED(arg);
+	u32_t dticks;
+
+#ifdef DEBUG_RTOS_TIMER
+	GPIO_CTRL_REGS->CTRL_0015 ^= (1ul << 16);
+#endif
+
+	timer_clear_intr();
+
+	cycle_count += last_load;
+	dticks = (cycle_count - announced_cycles) / CYC_PER_TICK;
+	announced_cycles += dticks * CYC_PER_TICK;
+
+	z_clock_announce(TICKLESS ? dticks : 1);
+}
+
+int z_clock_driver_init(struct device *device)
+{
+	mchp_pcr_periph_slp_ctrl(PCR_RTMR, 0);
+	mchp_pcr_periph_reset(PCR_RTMR);
+	timer_clear_intr();
+
+	last_load = CYC_PER_TICK;
+	cycle_count = 0U;
+	announced_cycles = 0U;
+	elapsed_counts = 0U;
+
+	IRQ_CONNECT(RTMR_IRQn, 1, xec_rtos_timer_isr, 0, 0);
+	irq_enable(RTMR_IRQn);
```

Comment:
Might consider moving irq_enable() to the very end of the process to avoid receiving spurious interrupts from a device in unknown state.